### PR TITLE
Bug 2004542: UPSTREAM: 89885: SQUASH: Retry fetching clouds.conf

### PR DESCRIPTION
**Pull request by pierreprinetti**
The OpenStack secret is not guaranteed to be present at the time
kube-controller-manager is initialised.

### CODE REVIEW

**Comment by mandre** [comment]
The patch is working as expected. From the kube-controller-manager logs in the must-gather:
```
2021-12-22T18:05:52.607911397Z W1222 18:05:52.607586       1 plugins.go:132] WARNING: openstack built-in cloud provider is now deprecated. Please use 'external' cloud provider for openstack: https://github.com/kubernetes/cloud-provider-openstack
2021-12-22T18:05:52.630425734Z I1222 18:05:52.629847       1 shared_informer.go:240] Waiting for caches to sync for tokens
2021-12-22T18:05:52.630425734Z I1222 18:05:52.630181       1 openstack.go:289] failed to get cloud.conf from secret (try 1), retrying after 1 seconds. Error was: secret "openstack-credentials" not found
2021-12-22T18:05:53.031045194Z I1222 18:05:53.030756       1 shared_informer.go:247] Caches are synced for tokens 
2021-12-22T18:05:55.061025884Z I1222 18:05:55.060964       1 openstack.go:313] OpenStack cloud provider was initialized using data from the secret.
```
The secret was missing on the first try, but was successfully retrieved on the second try, a second later.

**Comment by mandre** [comment]
To give a little bit more of context, this PR changes a downstream only behavior (fetching the cloud provider config from a secret) to fix a race condition, where the secret might not be available at the time we initialize the cloud provider.

We'd like to backport this change back to 4.9 at a minimum, preferably 4.8 as a customer is hitting the bug there too.

We'll also need to squash it with https://github.com/openshift/kubernetes/commit/e0840d63195d298549528b649cb63e09b681af92 going forward.

**Comment by mandre** [comment]
/retitle Bug 2004542: UPSTREAM: 89885: SQUASH: Retry fetching clouds.conf

**Comment by mandre** [comment]
Waiting for https://github.com/openshift/kubernetes/pull/1087 to land.

**Comment by pierreprinetti** [comment]
Rebased onto #1087. This is now ready to merge.

/cc mandre
/assign soltysh

**Review by mandre — COMMENTED** [review]
/lgtm
/retest-required

File: staging/src/k8s.io/legacy-cloud-providers/openstack/openstack.go [hunk 1]
```diff
@@ -280,7 +280,18 @@ func (os *OpenStack) setConfigFromSecret() error {
...
 
-	secret, err := os.secretLister.Secrets(os.secretNamespace).Get(os.secretName)
+	var secret *v1.Secret
+	var err error
+
+	for i := 1; i <= 10; i++ {
+		secret, err = os.secretLister.Secrets(os.secretNamespace).Get(os.secretName)
+		if err != nil {
+			klog.Infof("failed to get cloud.conf from secret (try %d), retrying after %d seconds. Error was: %v", i, i, err)
+			time.Sleep(time.Duration(i) * time.Second)
```
[the remark this answers is not in the document]
Comment (pierreprinetti) [reply staging/src/k8s.io/legacy-cloud-providers/openstack/openstack.go:290]: Fine, I'll do. May I ask what is "wrong" with directly calling Sleep?

[the remark this answers is not in the document]
Comment (pierreprinetti) [reply staging/src/k8s.io/legacy-cloud-providers/openstack/openstack.go]: I'm using retry and wait now, I hope it's more in line with what you expected. PTAL

[the remark this answers is not in the document]
Comment (pierreprinetti) [reply staging/src/k8s.io/legacy-cloud-providers/openstack/openstack.go]: @soltysh ?

**Review by soltysh — APPROVED** [review]
/lgtm
/approve

**Pull request merged**